In [2]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import rpy2.robjects as robjects
from rpy2.robjects import pandas2ri



In [3]:
# 1. Load the data
patent_rel_2015_path = '/home/bld/bld_data/yuanjian/llama-trl/KG_data/patent/2015.csv'
patent_rel_2016_path = '/home/bld/bld_data/yuanjian/llama-trl/KG_data/patent/2016.csv'
patent_rel_2017_path = '/home/bld/bld_data/yuanjian/llama-trl/KG_data/patent/2017.csv'
patent_rel_2018_path = '/home/bld/bld_data/yuanjian/llama-trl/KG_data/patent/2018.csv'
patent_rel_2019_path = '/home/bld/bld_data/yuanjian/llama-trl/KG_data/patent/2019.csv'
patent_rel_2020_path = '/home/bld/bld_data/yuanjian/llama-trl/KG_data/patent/2020.csv'
patent_rel_2021_path = '/home/bld/bld_data/yuanjian/llama-trl/KG_data/patent/2021.csv'

In [4]:
patent_rel_2015 = pd.read_csv(patent_rel_2015_path)
patent_rel_2016 = pd.read_csv(patent_rel_2016_path)
patent_rel_2017 = pd.read_csv(patent_rel_2017_path)
patent_rel_2018 = pd.read_csv(patent_rel_2018_path)
patent_rel_2019 = pd.read_csv(patent_rel_2019_path)
patent_rel_2020 = pd.read_csv(patent_rel_2020_path)
patent_rel_2021 = pd.read_csv(patent_rel_2021_path)

In [5]:
patent_rel_2015

Unnamed: 0      ORCL      MSFT       HON       EMC      AKRX      RGEN  \
0          ORCL  1.000000  0.986089  0.510183  0.974666  0.001629  0.002580   
1          MSFT  0.986089  1.000000  0.516407  0.944854  0.008874  0.006731   
2           HON  0.510183  0.516407  1.000000  0.432178  0.057296  0.224675   
3           EMC  0.974666  0.944854  0.432178  1.000000  0.001722  0.001295   
4          AKRX  0.001629  0.008874  0.057296  0.001722  1.000000  0.139382   
...         ...       ...       ...       ...       ...       ...       ...   
1377       CODE  0.218175  0.144038  0.393262  0.101333  0.001635  0.000000   
1378       FURX  0.000579  0.007384  0.050794  0.001383  0.101333  0.002125   
1379       TNAV  0.444101  0.466311  0.725561  0.050794  0.393262  0.099497   
1380       BSFT  0.395881  0.512083  0.466311  0.007384  0.144038  0.004396   
1381       TSLA  0.094610  0.395881  0.444101  0.000579  0.218175  0.002870   

          BEAM      CLDX      SIGM  ...      AOSL      CDXS      MITL  \
0     0.009070  0.000782  0.555487  ...  0.112198  0.001026  0.497179   
1     0.011793  0.007465  0.657208  ...  0.042425  0.001864  0.605568   
2     0.142817  0.107796  0.388204  ...  0.343717  0.037051  0.370293   
3     0.006619  0.001319  0.393424  ...  0.003209  0.000462  0.325612   
4     0.000000  0.893435  0.000000  ...  0.000190  0.044992  0.000566   
...        ...       ...       ...  ...       ...       ...       ...   
1377  0.000000  0.000566  0.044992  ...  0.013148  0.000000  0.893435   
1378  0.945075  0.325612  0.000462  ...  0.118742  0.393424  0.001319   
1379  0.357598  0.370293  0.037051  ...  0.392452  0.388204  0.107796   
1380  0.827872  0.605568  0.001864  ...  0.168531  0.657208  0.007465   
1381  0.891872  0.497179  0.001026  ...  0.197374  0.555487  0.000782   

           OCZ      PLOW      CODE      FURX      TNAV      BSFT      TSLA  
0     0.891872  0.002870  0.218175  0.000579  0.444101  0.395881  0.094610  
1     0.827872  0.004396  0.144038  0.007384  0.466311  0.512083  0.395881  
2     0.357598  0.099497  0.393262  0.050794  0.725561  0.466311  0.444101  
3     0.945075  0.002125  0.101333  0.001383  0.050794  0.007384  0.000579  
4     0.000000  0.000000  0.001635  0.101333  0.393262  0.144038  0.218175  
...        ...       ...       ...       ...       ...       ...       ...  
1377  0.000000  0.139382  1.000000  0.001722  0.057296  0.008874  0.001629  
1378  0.006619  0.001295  0.001722  1.000000  0.432178  0.944854  0.974666  
1379  0.142817  0.224675  0.057296  0.432178  1.000000  0.516407  0.510183  
1380  0.011793  0.006731  0.008874  0.944854  0.516407  1.000000  0.986089  
1381  0.009070  0.002580  0.001629  0.974666  0.510183  0.986089  1.000000  

[1382 rows x 1383 columns]

In [6]:
patent_rel_2016

Unnamed: 0      ORCL      MSFT       HON       EMC      AKRX      RGEN  \
0          ORCL  1.000000  0.989368  0.509508  0.978049  0.001576  0.001355   
1          MSFT  0.989368  1.000000  0.531334  0.949782  0.010126  0.005379   
2           HON  0.509508  0.531334  1.000000  0.437854  0.050743  0.173007   
3           EMC  0.978049  0.949782  0.437854  1.000000  0.001951  0.001006   
4          AKRX  0.001576  0.010126  0.050743  0.001951  1.000000  0.269389   
...         ...       ...       ...       ...       ...       ...       ...   
1332       CODE  0.192239  0.147525  0.359240  0.101795  0.001635  0.000000   
1333       FURX  0.000673  0.008646  0.044863  0.001567  0.101795  0.001956   
1334       TNAV  0.448759  0.477722  0.772542  0.044863  0.359240  0.090692   
1335       BSFT  0.398037  0.502544  0.477722  0.008646  0.147525  0.004157   
1336       TSLA  0.078631  0.398037  0.448759  0.000673  0.192239  0.002139   

          BEAM      CLDX      SIGM  ...      AOSL      CDXS      MITL  \
0     0.006958  0.000996  0.397688  ...  0.085966  0.000866  0.487008   
1     0.010450  0.008762  0.501468  ...  0.046491  0.001883  0.585600   
2     0.118539  0.119821  0.374519  ...  0.308342  0.032032  0.402222   
3     0.006145  0.001479  0.223343  ...  0.004019  0.000500  0.318354   
4     0.000000  0.852376  0.003807  ...  0.000228  0.058606  0.000497   
...        ...       ...       ...  ...       ...       ...       ...   
1332  0.000000  0.000497  0.058606  ...  0.015749  0.003807  0.852376   
1333  0.943708  0.318354  0.000500  ...  0.134962  0.223343  0.001479   
1334  0.352765  0.402222  0.032032  ...  0.329003  0.374519  0.119821   
1335  0.832813  0.585600  0.001883  ...  0.182224  0.501468  0.008762   
1336  0.893659  0.487008  0.000866  ...  0.183324  0.397688  0.000996   

           OCZ      PLOW      CODE      FURX      TNAV      BSFT      TSLA  
0     0.893659  0.002139  0.192239  0.000673  0.448759  0.398037  0.078631  
1     0.832813  0.004157  0.147525  0.008646  0.477722  0.502544  0.398037  
2     0.352765  0.090692  0.359240  0.044863  0.772542  0.477722  0.448759  
3     0.943708  0.001956  0.101795  0.001567  0.044863  0.008646  0.000673  
4     0.000000  0.000000  0.001635  0.101795  0.359240  0.147525  0.192239  
...        ...       ...       ...       ...       ...       ...       ...  
1332  0.000000  0.269389  1.000000  0.001951  0.050743  0.010126  0.001576  
1333  0.006145  0.001006  0.001951  1.000000  0.437854  0.949782  0.978049  
1334  0.118539  0.173007  0.050743  0.437854  1.000000  0.531334  0.509508  
1335  0.010450  0.005379  0.010126  0.949782  0.531334  1.000000  0.989368  
1336  0.006958  0.001355  0.001576  0.978049  0.509508  0.989368  1.000000  

[1337 rows x 1338 columns]

In [7]:
patent_rel_2017

Unnamed: 0      ORCL      MSFT       HON       EMC      AKRX      RGEN  \
0          ORCL  1.000000  0.994465  0.519483  0.977091  0.001078  0.000861   
1          MSFT  0.994465  1.000000  0.543269  0.958482  0.011693  0.004567   
2           HON  0.519483  0.543269  1.000000  0.451427  0.039484  0.148556   
3           EMC  0.977091  0.958482  0.451427  1.000000  0.001861  0.000624   
4          AKRX  0.001078  0.011693  0.039484  0.001861  1.000000  0.190931   
...         ...       ...       ...       ...       ...       ...       ...   
1291       CODE  0.189557  0.171537  0.363990  0.130693  0.000515  0.000000   
1292       FURX  0.000567  0.010803  0.030378  0.001616  0.130693  0.001621   
1293       TNAV  0.462799  0.486557  0.788666  0.030378  0.363990  0.099006   
1294       BSFT  0.428457  0.488123  0.486557  0.010803  0.171537  0.003147   
1295       TSLA  0.077360  0.428457  0.462799  0.000567  0.189557  0.000772   

          BEAM      CLDX      SIGM  ...      MERU      AOSL      CDXS  \
0     0.003732  0.001024  0.427946  ...  0.428457  0.059620  0.000742   
1     0.006631  0.010831  0.486832  ...  0.488123  0.047055  0.002402   
2     0.099254  0.118050  0.389570  ...  0.376053  0.303643  0.025020   
3     0.005742  0.001423  0.238434  ...  0.238754  0.004598  0.000491   
4     0.000000  0.830019  0.002730  ...  0.000000  0.000149  0.061915   
...        ...       ...       ...  ...       ...       ...       ...   
1291  0.000431  0.061915  0.000149  ...  0.010116  0.002730  0.830019   
1292  0.339046  0.000491  0.004598  ...  0.139516  0.238434  0.001423   
1293  0.420385  0.025020  0.303643  ...  0.309510  0.389570  0.118050   
1294  0.577482  0.002402  0.047055  ...  0.188275  0.486832  0.010831   
1295  0.521317  0.000742  0.059620  ...  0.189042  0.427946  0.001024   

          MITL      PLOW      CODE      FURX      TNAV      BSFT      TSLA  
0     0.521317  0.000772  0.189557  0.000567  0.462799  0.428457  0.077360  
1     0.577482  0.003147  0.171537  0.010803  0.486557  0.488123  0.428457  
2     0.420385  0.099006  0.363990  0.030378  0.788666  0.486557  0.462799  
3     0.339046  0.001621  0.130693  0.001616  0.030378  0.010803  0.000567  
4     0.000431  0.000000  0.000515  0.130693  0.363990  0.171537  0.189557  
...        ...       ...       ...       ...       ...       ...       ...  
1291  0.000000  0.190931  1.000000  0.001861  0.039484  0.011693  0.001078  
1292  0.005742  0.000624  0.001861  1.000000  0.451427  0.958482  0.977091  
1293  0.099254  0.148556  0.039484  0.451427  1.000000  0.543269  0.519483  
1294  0.006631  0.004567  0.011693  0.958482  0.543269  1.000000  0.994465  
1295  0.003732  0.000861  0.001078  0.977091  0.519483  0.994465  1.000000  

[1296 rows x 1297 columns]

In [8]:
patent_rel_2018

Unnamed: 0      ORCL      MSFT       HON       EMC      AKRX      RGEN  \
0          ORCL  1.000000  0.995748  0.523406  0.977833  0.000935  0.001170   
1          MSFT  0.995748  1.000000  0.548225  0.961634  0.012401  0.005284   
2           HON  0.523406  0.548225  1.000000  0.455731  0.040651  0.151989   
3           EMC  0.977833  0.961634  0.455731  1.000000  0.000301  0.000165   
4          AKRX  0.000935  0.012401  0.040651  0.000301  1.000000  0.209336   
...         ...       ...       ...       ...       ...       ...       ...   
1229       CODE  0.128431  0.123486  0.332293  0.077414  0.000000  0.000000   
1230       FURX  0.000463  0.011573  0.031515  0.000000  0.077414  0.000702   
1231       TNAV  0.439585  0.462896  0.800408  0.031515  0.332293  0.088491   
1232       BSFT  0.426020  0.475925  0.462896  0.011573  0.123486  0.002264   
1233       TSLA  0.048117  0.426020  0.439585  0.000463  0.128431  0.000918   

          CLDX      SIGM      LLTC  ...      MERU      AOSL      CDXS  \
0     0.001241  0.425638  0.184068  ...  0.426020  0.044347  0.000664   
1     0.011061  0.475831  0.183248  ...  0.475925  0.038689  0.002574   
2     0.143828  0.406951  0.300477  ...  0.391569  0.283595  0.022037   
3     0.000089  0.233823  0.142576  ...  0.233974  0.003263  0.000438   
4     0.695205  0.003349  0.008540  ...  0.000000  0.000157  0.083400   
...        ...       ...       ...  ...       ...       ...       ...   
1229  0.000598  0.083400  0.000157  ...  0.010533  0.008540  0.003349   
1230  0.360281  0.000438  0.003263  ...  0.470253  0.142576  0.233823   
1231  0.446942  0.022037  0.283595  ...  0.487271  0.300477  0.406951   
1232  0.588551  0.002574  0.038689  ...  0.479662  0.183248  0.475831   
1233  0.542536  0.000664  0.044347  ...  0.496500  0.184068  0.425638   

          MITL      PLOW      CODE      FURX      TNAV      BSFT      TSLA  
0     0.542536  0.000918  0.128431  0.000463  0.439585  0.426020  0.048117  
1     0.588551  0.002264  0.123486  0.011573  0.462896  0.475925  0.426020  
2     0.446942  0.088491  0.332293  0.031515  0.800408  0.462896  0.439585  
3     0.360281  0.000702  0.077414  0.000000  0.031515  0.011573  0.000463  
4     0.000598  0.000000  0.000000  0.077414  0.332293  0.123486  0.128431  
...        ...       ...       ...       ...       ...       ...       ...  
1229  0.695205  0.209336  1.000000  0.000301  0.040651  0.012401  0.000935  
1230  0.000089  0.000165  0.000301  1.000000  0.455731  0.961634  0.977833  
1231  0.143828  0.151989  0.040651  0.455731  1.000000  0.548225  0.523406  
1232  0.011061  0.005284  0.012401  0.961634  0.548225  1.000000  0.995748  
1233  0.001241  0.001170  0.000935  0.977833  0.523406  0.995748  1.000000  

[1234 rows x 1235 columns]

In [9]:
patent_rel_2019

Unnamed: 0      ORCL      MSFT       HON       EMC      AKRX      RGEN  \
0          ORCL  1.000000  0.997008  0.538021  0.960023  0.000978  0.001665   
1          MSFT  0.997008  1.000000  0.560885  0.950699  0.012195  0.006453   
2           HON  0.538021  0.560885  1.000000  0.455761  0.042815  0.177261   
3           EMC  0.960023  0.950699  0.455761  1.000000  0.000185  0.000822   
4          AKRX  0.000978  0.012195  0.042815  0.000185  1.000000  0.162403   
...         ...       ...       ...       ...       ...       ...       ...   
1150       PLOW  0.000650  0.001864  0.090537  0.001009  0.000000  0.000000   
1151       CODE  0.203449  0.220088  0.408852  0.079430  0.001009  0.321796   
1152       TNAV  0.432229  0.450995  0.823979  0.408852  0.090537  0.451920   
1153       BSFT  0.457431  0.473063  0.450995  0.220088  0.001864  0.586426   
1154       TSLA  0.084628  0.457431  0.432229  0.203449  0.000650  0.572018   

          CLDX      SIGM      LLTC  ...      AVEO      ALIM      AOSL  \
0     0.001091  0.456758  0.183497  ...  0.000184  0.000391  0.026394   
1     0.010029  0.472786  0.185196  ...  0.005523  0.011039  0.037886   
2     0.125156  0.421664  0.272581  ...  0.076502  0.029565  0.274589   
3     0.000243  0.192169  0.151125  ...  0.000000  0.000000  0.005024   
4     0.665385  0.006137  0.010375  ...  0.464740  0.989949  0.000245   
...        ...       ...       ...  ...       ...       ...       ...   
1150  0.120006  0.000245  0.989949  ...  0.014506  0.010375  0.006137   
1151  0.000207  0.005024  0.000000  ...  0.430725  0.151125  0.192169   
1152  0.019303  0.274589  0.029565  ...  0.481543  0.272581  0.421664   
1153  0.002866  0.037886  0.011039  ...  0.450485  0.185196  0.472786   
1154  0.000511  0.026394  0.000391  ...  0.446699  0.183497  0.456758   

          CDXS      MITL      PLOW      CODE      TNAV      BSFT      TSLA  
0     0.000511  0.572018  0.000650  0.203449  0.432229  0.457431  0.084628  
1     0.002866  0.586426  0.001864  0.220088  0.450995  0.473063  0.457431  
2     0.019303  0.451920  0.090537  0.408852  0.823979  0.450995  0.432229  
3     0.000207  0.321796  0.001009  0.079430  0.408852  0.220088  0.203449  
4     0.120006  0.000000  0.000000  0.001009  0.090537  0.001864  0.000650  
...        ...       ...       ...       ...       ...       ...       ...  
1150  0.665385  0.162403  1.000000  0.000185  0.042815  0.012195  0.000978  
1151  0.000243  0.000822  0.000185  1.000000  0.455761  0.950699  0.960023  
1152  0.125156  0.177261  0.042815  0.455761  1.000000  0.560885  0.538021  
1153  0.010029  0.006453  0.012195  0.950699  0.560885  1.000000  0.997008  
1154  0.001091  0.001665  0.000978  0.960023  0.538021  0.997008  1.000000  

[1155 rows x 1156 columns]

In [10]:
patent_rel_2020

Unnamed: 0      ORCL      MSFT       HON       EMC      AKRX      RGEN  \
0          ORCL  1.000000  0.996849  0.543983  0.959099  0.000701  0.001290   
1          MSFT  0.996849  1.000000  0.568427  0.950150  0.010747  0.005444   
2           HON  0.543983  0.568427  1.000000  0.459842  0.040836  0.160844   
3           EMC  0.959099  0.950150  0.459842  1.000000  0.000000  0.000000   
4          AKRX  0.000701  0.010747  0.040836  0.000000  1.000000  0.144885   
...         ...       ...       ...       ...       ...       ...       ...   
1068       CDXS  0.000376  0.002792  0.016586  0.000048  0.133591  0.000250   
1069       MITL  0.596193  0.607721  0.473763  0.346441  0.000048  0.011325   
1070       PLOW  0.000700  0.001906  0.072921  0.473763  0.016586  0.257662   
1071       TNAV  0.382072  0.402273  0.001906  0.607721  0.002792  0.037004   
1072       TSLA  0.147936  0.382072  0.000700  0.596193  0.000376  0.022413   

          CLDX      SIGM      LLTC  ...       SPB        ST      AVEO  \
0     0.000679  0.450340  0.130836  ...  0.035798  0.096296  0.000094   
1     0.008773  0.464313  0.140147  ...  0.049440  0.120833  0.004874   
2     0.109382  0.431209  0.293978  ...  0.192397  0.696354  0.062311   
3     0.000000  0.181474  0.099468  ...  0.017599  0.050859  0.000000   
4     0.657190  0.006137  0.006683  ...  0.011173  0.003806  0.464740   
...        ...       ...       ...  ...       ...       ...       ...   
1068  0.989949  0.464740  0.003806  ...  0.020192  0.006683  0.006137   
1069  0.000000  0.000000  0.050859  ...  0.520381  0.099468  0.181474   
1070  0.028113  0.062311  0.696354  ...  0.506136  0.293978  0.431209   
1071  0.009600  0.004874  0.120833  ...  0.541585  0.140147  0.464313   
1072  0.000201  0.000094  0.096296  ...  0.534894  0.130836  0.450340   

          ALIM      AOSL      CDXS      MITL      PLOW      TNAV      TSLA  
0     0.000201  0.022413  0.000376  0.596193  0.000700  0.382072  0.147936  
1     0.009600  0.037004  0.002792  0.607721  0.001906  0.402273  0.382072  
2     0.028113  0.257662  0.016586  0.473763  0.072921  0.001906  0.000700  
3     0.000000  0.011325  0.000048  0.346441  0.473763  0.607721  0.596193  
4     0.989949  0.000250  0.133591  0.000048  0.016586  0.002792  0.000376  
...        ...       ...       ...       ...       ...       ...       ...  
1068  0.657190  0.144885  1.000000  0.000000  0.040836  0.010747  0.000701  
1069  0.000000  0.000000  0.000000  1.000000  0.459842  0.950150  0.959099  
1070  0.109382  0.160844  0.040836  0.459842  1.000000  0.568427  0.543983  
1071  0.008773  0.005444  0.010747  0.950150  0.568427  1.000000  0.996849  
1072  0.000679  0.001290  0.000701  0.959099  0.543983  0.996849  1.000000  

[1073 rows x 1074 columns]

In [11]:
patent_rel_2021

Unnamed: 0      ORCL      MSFT       HON      AKRX      RGEN      CLDX  \
0         ORCL  1.000000  0.996954  0.544004  0.000237  0.001480  0.000715   
1         MSFT  0.996954  1.000000  0.567932  0.010038  0.005946  0.008146   
2          HON  0.544004  0.567932  1.000000  0.025809  0.159239  0.109113   
3         AKRX  0.000237  0.010038  0.025809  1.000000  0.134283  0.562228   
4         RGEN  0.001480  0.005946  0.159239  0.134283  1.000000  0.148286   
..         ...       ...       ...       ...       ...       ...       ...   
962       CDXS  0.000366  0.003072  0.014435  0.143459  0.442706  0.000063   
963       MITL  0.521793  0.528116  0.439144  0.000000  0.143459  0.000000   
964       PLOW  0.000759  0.001396  0.074570  0.439144  0.014435  0.226870   
965       TNAV  0.369116  0.382997  0.001396  0.528116  0.003072  0.036722   
966       TSLA  0.005866  0.369116  0.000759  0.521793  0.000366  0.018454   

         SIGM        CY      BCPC  ...       SPB        ST      AVEO  \
0    0.453124  0.535537  0.000233  ...  0.051475  0.107827  0.000094   
1    0.460487  0.547536  0.009883  ...  0.067724  0.132626  0.004398   
2    0.428693  0.504312  0.031441  ...  0.229564  0.697259  0.059364   
3    0.000000  0.021734  0.981981  ...  0.000000  0.000000  0.397894   
4    0.000971  0.010418  0.138376  ...  0.033267  0.111314  0.214783   
..        ...       ...       ...  ...       ...       ...       ...   
962  0.134283  0.214783  0.111314  ...  0.146369  0.138376  0.010418   
963  1.000000  0.397894  0.000000  ...  0.946260  0.981981  0.021734   
964  0.025809  0.059364  0.697259  ...  0.042422  0.031441  0.504312   
965  0.010038  0.004398  0.132626  ...  0.009573  0.009883  0.547536   
966  0.000237  0.000094  0.107827  ...  0.000224  0.000233  0.535537   

         ALIM      AOSL      CDXS      MITL      PLOW      TNAV      TSLA  
0    0.000237  0.018454  0.000366  0.521793  0.000759  0.369116  0.005866  
1    0.010038  0.036722  0.003072  0.528116  0.001396  0.382997  0.369116  
2    0.025809  0.226870  0.014435  0.439144  0.074570  0.001396  0.000759  
3    1.000000  0.000000  0.143459  0.000000  0.439144  0.528116  0.521793  
4    0.134283  0.000063  0.442706  0.143459  0.014435  0.003072  0.000366  
..        ...       ...       ...       ...       ...       ...       ...  
962  0.000971  0.148286  1.000000  0.134283  0.159239  0.005946  0.001480  
963  0.000000  0.562228  0.134283  1.000000  0.025809  0.010038  0.000237  
964  0.428693  0.109113  0.159239  0.025809  1.000000  0.567932  0.544004  
965  0.460487  0.008146  0.005946  0.010038  0.567932  1.000000  0.996954  
966  0.453124  0.000715  0.001480  0.000237  0.544004  0.996954  1.000000  

[967 rows x 968 columns]

In [12]:
# 逐个提取专利csv文件,从1965年到2021年
# 用for 循环提取
start_year = 1965
end_year = 2021
patent_rel_list = []
for year in range(start_year, end_year+1):
    patent_rel_path = '/home/bld/bld_data/yuanjian/llama-trl/KG_data/patent/{}.csv'.format(year)
    patent_rel = pd.read_csv(patent_rel_path)
    patent_rel_list.append(patent_rel)
    print('The shape of {} is {}'.format(year, patent_rel.shape))
    

The shape of 1965 is (49, 50)
The shape of 1966 is (55, 56)
The shape of 1967 is (67, 68)
The shape of 1968 is (73, 74)
The shape of 1969 is (96, 97)
The shape of 1970 is (129, 130)
The shape of 1971 is (185, 186)
The shape of 1972 is (254, 255)
The shape of 1973 is (382, 383)
The shape of 1974 is (645, 646)
The shape of 1975 is (787, 788)
The shape of 1976 is (867, 868)
The shape of 1977 is (907, 908)
The shape of 1978 is (941, 942)
The shape of 1979 is (959, 960)
The shape of 1980 is (953, 954)
The shape of 1981 is (923, 924)
The shape of 1982 is (878, 879)
The shape of 1983 is (826, 827)
The shape of 1984 is (786, 787)
The shape of 1985 is (778, 779)
The shape of 1986 is (773, 774)
The shape of 1987 is (791, 792)
The shape of 1988 is (842, 843)
The shape of 1989 is (858, 859)
The shape of 1990 is (865, 866)
The shape of 1991 is (884, 885)
The shape of 1992 is (947, 948)
The shape of 1993 is (1046, 1047)
The shape of 1994 is (1174, 1175)
The shape of 1995 is (1313, 1314)
The shape of

In [13]:
patent_rel_list[15]

Unnamed: 0       ACF       AMT        AN       HON        AH       AMX  \
0          ACF  1.000000  0.386795  0.476938  0.410143  0.825841  0.072133   
1          AMT  0.386795  1.000000  0.513314  0.450839  0.358642  0.076719   
2           AN  0.476938  0.513314  1.000000  0.491018  0.444663  0.200798   
3          HON  0.410143  0.450839  0.491018  1.000000  0.399225  0.291340   
4           AH  0.825841  0.358642  0.444663  0.399225  1.000000  0.106478   
..         ...       ...       ...       ...       ...       ...       ...   
948        GSE  0.290809  0.195316  0.252183  0.282104  0.287334  0.045284   
949        PGO  0.019455  0.179888  0.000000  0.006806  0.000000  0.008255   
950      MMBLF  0.334525  0.285342  0.277358  0.284720  0.248426  0.047922   
951      MALTZ  0.000000  0.000000  0.036202  0.016910  0.000000  0.000000   
952        WIX  0.500673  0.731704  0.564469  0.403137  0.380131  0.063472   

           AB      BEAM        PA  ...       NWP       EMH       FCT  \
0    0.303656  0.146780  0.512033  ...  0.014501  0.455611  0.842134   
1    0.064602  0.155758  0.475792  ...  0.000000  0.485179  0.192410   
2    0.040253  0.127007  0.418169  ...  0.036202  0.526994  0.359604   
3    0.148530  0.210524  0.598842  ...  0.109912  0.504015  0.278867   
4    0.108736  0.138071  0.392914  ...  0.050450  0.388469  0.903841   
..        ...       ...       ...  ...       ...       ...       ...   
948  0.126651  0.069216  0.239333  ...  0.569540  0.312474  0.240674   
949  0.000000  0.017317  0.007875  ...  0.000000  0.000000  0.000000   
950  0.000000  0.107459  0.417441  ...  0.000000  0.386022  0.171785   
951  0.000000  0.000000  0.015407  ...  0.000000  0.050809  0.000000   
952  0.009499  0.161113  0.511104  ...  0.042718  0.472347  0.268195   

          LBP     WKW.1       GSE       PGO     MMBLF     MALTZ       WIX  
0    0.002342  0.000000  0.290809  0.019455  0.334525  0.000000  0.500673  
1    0.023199  0.000000  0.195316  0.179888  0.285342  0.000000  0.731704  
2    0.014198  0.506834  0.252183  0.000000  0.277358  0.036202  0.564469  
3    0.016267  0.010146  0.282104  0.006806  0.284720  0.016910  0.403137  
4    0.008237  0.000000  0.287334  0.000000  0.248426  0.000000  0.380131  
..        ...       ...       ...       ...       ...       ...       ...  
948  0.075802  0.000000  1.000000  0.117557  0.156636  0.000000  0.239553  
949  0.000000  0.000000  0.117557  1.000000  0.000000  0.000000  0.000000  
950  0.001964  0.000000  0.156636  0.000000  1.000000  0.000000  0.385457  
951  0.000000  0.000000  0.000000  0.000000  0.000000  1.000000  0.000000  
952  0.000000  0.000000  0.239553  0.000000  0.385457  0.000000  1.000000  

[953 rows x 954 columns]

In [14]:
# 逐个检查list中的数据是否存在空置
for i in range(len(patent_rel_list)):
    print('The number of missing values in {} is {}'.format(i+1965, patent_rel_list[i].isnull().sum().sum()))

The number of missing values in 1965 is 0
The number of missing values in 1966 is 0
The number of missing values in 1967 is 0
The number of missing values in 1968 is 0
The number of missing values in 1969 is 0
The number of missing values in 1970 is 0
The number of missing values in 1971 is 0
The number of missing values in 1972 is 507
The number of missing values in 1973 is 0
The number of missing values in 1974 is 5144
The number of missing values in 1975 is 3144
The number of missing values in 1976 is 8645
The number of missing values in 1977 is 9045
The number of missing values in 1978 is 9385
The number of missing values in 1979 is 11472
The number of missing values in 1980 is 7608
The number of missing values in 1981 is 3688
The number of missing values in 1982 is 0
The number of missing values in 1983 is 1651
The number of missing values in 1984 is 4707
The number of missing values in 1985 is 3108
The number of missing values in 1986 is 1545
The number of missing values in 1987 

In [15]:
# 打印1965-2021年数据中的都有的列名的交集
col_list = []
for i in range(45,len(patent_rel_list)):
    col_list.append(patent_rel_list[i].columns)
col_list_set = set(col_list[0])
for i in range(1, len(col_list)):
    col_list_set = col_list_set.intersection(set(col_list[i]))
col_list_set



{'A',
 'AAL',
 'AATC',
 'ABB',
 'ABIO',
 'ABMD',
 'ACAD',
 'ACCO',
 'ACHN',
 'ACLS',
 'ACN',
 'ACOR',
 'ADES',
 'ADM',
 'ADTN',
 'AEHR',
 'AEIS',
 'AES',
 'AET',
 'AG',
 'AGEN',
 'AIZ',
 'AKAM',
 'ALB',
 'ALG',
 'ALKS',
 'ALL',
 'ALLT',
 'ALNY',
 'ALV',
 'ALXN',
 'AMAG',
 'AMCC',
 'AMGN',
 'AMKR',
 'AMP',
 'AMRI',
 'AMSC',
 'AMZN',
 'ANDE',
 'ANGO',
 'AOS',
 'APA',
 'APD',
 'APH',
 'APOL',
 'APYX',
 'ARAY',
 'ARLZ',
 'ARNA',
 'ARQL',
 'ARRS',
 'ARRY',
 'ARW',
 'ASML',
 'ASRT',
 'ASTE',
 'ASX',
 'ATHE',
 'ATHX',
 'ATR',
 'ATRC',
 'ATRS',
 'ATVI',
 'AUDC',
 'AUO',
 'AVGO',
 'AVID',
 'AVNT',
 'AVP',
 'AVY',
 'AWH',
 'AWK',
 'AWRE',
 'AXDX',
 'AXL',
 'AXON',
 'AYI',
 'AZN',
 'AZPN',
 'AZTA',
 'B',
 'BA',
 'BAX',
 'BB',
 'BBY',
 'BCR',
 'BCRX',
 'BDC',
 'BDSI',
 'BDX',
 'BEAV',
 'BGCP',
 'BIDU',
 'BIIB',
 'BK',
 'BKNG',
 'BKR',
 'BLK',
 'BMI',
 'BMRN',
 'BMS',
 'BMY',
 'BPA',
 'BRCD',
 'BRKR',
 'BSX',
 'BWA',
 'CAMT',
 'CAT',
 'CAVM',
 'CBIO',
 'CBT',
 'CDNS',
 'CDTI',
 'CELG',
 'CEVA',
 'C

In [16]:
len(col_list_set)

632

## Existing issues:
-  不同年份文件中的公司数量不一致
-  2018-2021年的数据存在空值（patent相似度计算通常来所不可能出现空值，需询问专利相似度计算过程和关系构建过程及原始文件）
-  

In [17]:
patent_rel_2015.isna().sum().sum()

0

In [18]:
patent_rel_2016.isna().sum().sum()

0

In [19]:
patent_rel_2017.isna().sum().sum()

0

In [20]:
patent_rel_2018.isna().sum().sum()

2508

In [21]:
patent_rel_2019.isna().sum().sum()

2270

In [22]:
patent_rel_2020.isna().sum().sum()

2094

In [23]:
patent_rel_2021.isna().sum().sum()

1838

In [24]:
null_positions = patent_rel_2018['MSFT'].isnull()
null_positions = null_positions[null_positions].index
print(null_positions)

Index([606], dtype='int64')


In [25]:
patent_rel_2018['MSFT'][605:610]

605    0.011573
606         NaN
607    0.004326
608    0.030785
609    0.745348
Name: MSFT, dtype: float64

In [26]:
patent_rel_2018['MSFT']

0       0.995748
1       1.000000
2       0.548225
3       0.961634
4       0.012401
          ...   
1229    0.123486
1230    0.011573
1231    0.462896
1232    0.475925
1233    0.426020
Name: MSFT, Length: 1234, dtype: float64

In [27]:
patent_rel_2018.iloc[605:610, :]

Unnamed: 0      ORCL      MSFT       HON       EMC      AKRX      RGEN  \
605       ANIK  0.000463  0.011573  0.031515  0.000000  0.995893  0.210199   
606        TWI       NaN       NaN       NaN       NaN       NaN       NaN   
607        USG  0.002086  0.004326  0.136310  0.000694  0.006022  0.062936   
608       THRM  0.031291  0.030785  0.218308  0.010958  0.053512  0.019756   
609       FLIR  0.710436  0.745348  0.803026  0.597648  0.010479  0.051847   

         CLDX      SIGM      LLTC  ...      MERU      AOSL      CDXS  \
605  0.698072  0.000000  0.000000  ...  0.723569  0.312891  0.454575   
606       NaN       NaN       NaN  ...  0.000000  0.000000  0.000000   
607  0.008009  0.002460  0.011574  ...  0.210384  0.217424  0.918513   
608  0.031091  0.004685  0.205137  ...  0.463462  0.127353  0.223924   
609  0.065131  0.674845  0.254386  ...  0.000000  0.000000  0.000000   

         MITL      PLOW      CODE      FURX      TNAV      BSFT      TSLA  
605  0.021568  0.016416  0.000000  0.121021  0.542331  0.280273  0.247615  
606  0.000000  0.000000  0.000000  0.000000  0.003558  0.000094  0.000000  
607  0.035028  0.026985  0.009485  0.387870  0.640092  0.602614  0.555495  
608  0.000000  0.001746  0.000000  0.997285  0.453699  0.956467  0.972874  
609  0.000000  0.000000  0.000000  0.000000  0.009907  0.000253  0.000574  

[5 rows x 1235 columns]

In [28]:
patent_rel_2019.iloc[605:610, :]

Unnamed: 0      ORCL      MSFT       HON       EMC      AKRX      RGEN  \
605         IT  0.970893  0.962785  0.471175  0.995185  0.000000  0.003711   
606       ITRI  0.531525  0.550916  0.605488  0.289347  0.013767  0.038393   
607       UFPI  0.000000  0.000091  0.002312  0.000000  0.000000  0.000000   
608       VIAV  0.361541  0.384460  0.547488  0.149299  0.000000  0.027450   
609       INCY  0.001599  0.011860  0.059276  0.001136  0.923174  0.178577   

         CLDX      SIGM      LLTC  ...      AVEO      ALIM      AOSL  \
605  0.000000  0.247720  0.132844  ...  0.178752  0.169649  0.985956   
606  0.026534  0.945291  0.194709  ...  0.081666  0.048059  0.039258   
607  0.000000  0.000000  0.000000  ...  0.234501  0.202469  0.459316   
608  0.020261  0.763460  0.245829  ...  0.001969  0.006377  0.002859   
609  0.893208  0.000717  0.000796  ...  0.001031  0.000000  0.000000   

         CDXS      MITL      PLOW      CODE      TNAV      BSFT      TSLA  
605  0.000000  0.000546  0.001913  0.322981  0.457727  0.590906  0.573275  
606  0.009272  0.044556  0.001607  0.157479  0.225390  0.158512  0.156041  
607  0.000723  0.002517  0.001102  0.435827  0.384381  0.581914  0.532683  
608  0.000000  0.008127  0.009317  0.000237  0.046386  0.002978  0.004110  
609  0.000000  0.398495  0.000000  0.000811  0.113252  0.000800  0.000515  

[5 rows x 1156 columns]

In [29]:
patent_rel_2018_colums = patent_rel_2018.columns[1:].to_list()
patent_rel_2018_colums

['ORCL',
 'MSFT',
 'HON',
 'EMC',
 'AKRX',
 'RGEN',
 'CLDX',
 'SIGM',
 'LLTC',
 'CY',
 'BCPC',
 'XOMA',
 'AMAG',
 'ASTE',
 'AMWD',
 'KSPN',
 'ADM',
 'VIVO',
 'WTS',
 'FISV',
 'RENT',
 'UIS',
 'PLAB',
 'KO',
 'CCC',
 'CDNS',
 'ED',
 'CNMD',
 'CELG',
 'EPAC',
 'XRAY',
 'HRTX',
 'GDOT',
 'DD',
 'IIVI',
 'ETN',
 'JOUT',
 'XOM',
 'MXIM',
 'RLD',
 'XSPA',
 'SMT',
 'GATX',
 'GE',
 'NXPI',
 'LNN',
 'AMRS',
 'PACB',
 'GM',
 'IPHI',
 'ANAC',
 'ABUS',
 'ZGNX',
 'TT',
 'IBM',
 'NPTN',
 'ITT',
 'ACRX',
 'ECYT',
 'FLDM',
 'GEVO',
 'PCRX',
 'WMGI',
 'USCR',
 'MX',
 'SREV',
 'ELLI',
 'PFLT',
 'OCRX',
 'LNKD',
 'FSL',
 'AMPE',
 'RPXC',
 'TVIA',
 'WIFI',
 'FIO',
 'P',
 'ATNY',
 'GRFS',
 'KIOR',
 'QTRH',
 'SKUL',
 'TNGO',
 'CARB',
 'FSM',
 'XLS',
 'UI',
 'ZLTQ',
 'IMPV',
 'INVN',
 'GRPN',
 'IMI',
 'LUMO',
 'JIVE',
 'SGYP',
 'ZNGA',
 'PRLB',
 'BCOV',
 'BV',
 'CCXI',
 'ALSN',
 'MM',
 'WP',
 'VCRA',
 'CSTE',
 'ENPH',
 'GALT',
 'MACK',
 'MTSI',
 'NVDQ',
 'HEAR',
 'FET',
 'GIG',
 'BLOX',
 'PSX',
 'CLIR',
 'PF

In [32]:
patent_original_data = pd.read_csv('/home/bld/bld_data/pengliu/Graph data/Patent relation/patent_Gen_new/patent_cpc_data.csv')
patent_original_data

patent_num                                                cpc cpc_1  \
0           1570604                                           F23K3/12     F   
1           1570665                    B26D3/16;Y10T82/16229;B26D3/003     B   
2           1570677                                F41A19/49;F41A19/15     F   
3           1570692      E05F5/025;Y10T16/628;Y10T16/6285;E05Y2900/531     E   
4           1570694                                          B60P3/073     B   
...             ...                                                ...   ...   
3160448    11540395  H05K2201/10159;H05K2201/10015;H01L23/5223;H01L...     H   
3160449    11540408  B23C2226/315;B23C2220/04;B23C2220/16;B23C2220/...     B   
3160450    11540421                  H05K7/20836;G06F1/206;H05K7/20436     H   
3160451    11540430  H05K5/0017;H05K7/1427;H01L33/642;H05K7/20963;H...     H   
3160452    11540432  H05K9/0064;B25J1/00;B65G49/07;H01J37/20;H01J37...     H   

        cpc_2  cpc_g1  cpc_g2 cpc_3  cpc_g3  permno  issue_date  filing_date  \
0         F23     6.0    99.0  F23K   480.0   15368    19260126   19230604.0   
1         B26     2.0    30.0  B26D   139.0   11965    19260126   19210627.0   
2         F41     6.0   105.0  F41A   508.0   10786    19260126   19220726.0   
3         E05     5.0    86.0  E05F   406.0   10807    19260126   19241117.0   
4         B60     2.0    42.0  B60P   201.0   14613    19260126   19231217.0   
...       ...     ...     ...   ...     ...     ...         ...          ...   
3160448   H05     8.0   126.0  H05K   656.0   59328    20221227   20190624.0   
3160449   B23     2.0    27.0  B23C   120.0   14593    20221227   20160722.0   
3160450   H05     8.0   126.0  H05K   656.0   89641    20221227   20200911.0   
3160451   H05     8.0   126.0  H05K   656.0   90266    20221227   20201225.0   
3160452   H05     8.0   126.0  H05K   656.0   14702    20221227   20200602.0   

         xi_nominal    xi_real  cites  year_issue  month_issue  year_filing  \
0          0.156578   0.863388      5        1926            1       1923.0   
1          0.137627   0.758891      2        1926            1       1921.0   
2          0.029288   0.161497      0        1926            1       1922.0   
3          0.508039   2.801388      0        1926            1       1924.0   
4          0.193691   1.068035      0        1926            1       1923.0   
...             ...        ...    ...         ...          ...          ...   
3160448   11.867290   3.902491      0        2022           12       2019.0   
3160449  168.119060  55.284992      0        2022           12       2016.0   
3160450   15.761281   5.183007      0        2022           12       2020.0   
3160451    0.010343   0.003401      0        2022           12       2020.0   
3160452   52.425938  17.239969      0        2022           12       2020.0   

         month_filing  
0                 6.0  
1                 6.0  
2                 7.0  
3                11.0  
4                12.0  
...               ...  
3160448           6.0  
3160449           7.0  
3160450           9.0  
3160451          12.0  
3160452           6.0  

[3160453 rows x 18 columns]

当然，以下是这些数据字段的中文解释：

1. **patent_num**: 专利编号
2. **cpc**: 国际专利分类 (Cooperative Patent Classification)
3. **cpc_1**: CPC 一级分类代码
4. **cpc_2**: CPC 二级分类代码
5. **cpc_g1**: CPC 一级分类的数量组
6. **cpc_g2**: CPC 二级分类的数量组
7. **cpc_3**: CPC 三级分类代码
8. **cpc_g3**: CPC 三级分类的数量组
9. **permno**: 永久编号
10. **issue_date**: 授权日期
11. **filing_date**: 申请日期
12. **xi_nominal**: 名义价值指数
13. **xi_real**: 实际价值指数
14. **cites**: 被引用次数
15. **year_issue**: 授权年份
16. **month_issue**: 授权月份
17. **year_filing**: 申请年份
18. **month_filing**: 申请月份


In [ ]:
patent_original_data['year_issue'].unique()

array([1926, 1927, 1928, 1929, 1930, 1931, 1932, 1933, 1934, 1935, 1936,
       1937, 1938, 1939, 1940, 1941, 1942, 1943, 1944, 1945, 1946, 1947,
       1948, 1949, 1950, 1951, 1952, 1953, 1954, 1955, 1956, 1957, 1958,
       1959, 1960, 1961, 1962, 1963, 1964, 1965, 1966, 1967, 1968, 1969,
       1970, 1971, 1972, 1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980,
       1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991,
       1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002,
       2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013,
       2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022])

In [ ]:
len(patent_original_data['year_issue'].unique())

97

In [ ]:
# 筛选出permno=10006的数据
patent_original_data_10006 = patent_original_data[patent_original_data['permno']==10006]
patent_original_data_10006

patent_num                                               cpc cpc_1  \
648        1579202                                          B61K5/00     B   
649        1579203  G01F23/04;Y10T137/8342;Y10T137/86324;Y10T137/863     G   
653        1579214                                         B60B17/00     B   
656        1579222              F16K27/07;Y10T137/8225;Y10T137/86316     F   
657        1579223             F16L27/0845;Y10T137/8807;Y10T137/5762     F   
...            ...                                               ...   ...   
802987     4447039            B61D5/08;F16K27/07;F16K51/00;B60P3/224     B   
804776     4452202                     F02D41/18;G01L9/02;G01L9/0057     F   
805146     4453192                               H02H3/253;H02H11/00     H   
805765     4455110                                 B61D7/02;B61D7/20     B   
806119     4456220                                         F16K27/07     F   

       cpc_2  cpc_g1  cpc_g2 cpc_3  cpc_g3  permno  issue_date  filing_date  \
648      B61     2.0    43.0  B61K   216.0   10006    19260406   19220517.0   
649      G01     7.0   107.0  G01F   521.0   10006    19260406   19241206.0   
653      B60     2.0    42.0  B60B   190.0   10006    19260406   19250814.0   
656      F16     6.0    95.0  F16K   454.0   10006    19260406   19210629.0   
657      F16     6.0    95.0  F16L   455.0   10006    19260406   19210909.0   
...      ...     ...     ...   ...     ...     ...         ...          ...   
802987   B61     2.0    43.0  B61D   211.0   10006    19840508   19830110.0   
804776   F02     6.0    90.0  F02D   425.0   10006    19840605   19811224.0   
805146   H02     8.0   123.0  H02H   622.0   10006    19840605   19810226.0   
805765   B61     2.0    43.0  B61D   211.0   10006    19840619   19820222.0   
806119   F16     6.0    95.0  F16K   454.0   10006    19840626   19810601.0   

        xi_nominal   xi_real  cites  year_issue  month_issue  year_filing  \
648       0.011230  0.061923      0        1926            4       1922.0   
649       0.011230  0.061923      0        1926            4       1924.0   
653       0.011230  0.061923      0        1926            4       1925.0   
656       0.011230  0.061923      2        1926            4       1921.0   
657       0.011230  0.061923      2        1926            4       1921.0   
...            ...       ...    ...         ...          ...          ...   
802987    3.234042  2.997555      0        1984            5       1983.0   
804776    1.542302  1.429522     24        1984            6       1981.0   
805146    1.542302  1.429522      3        1984            6       1981.0   
805765    3.020852  2.799954      2        1984            6       1982.0   
806119    3.174890  2.942728      1        1984            6       1981.0   

        month_filing  
648              5.0  
649             12.0  
653              8.0  
656              6.0  
657              9.0  
...              ...  
802987           1.0  
804776          12.0  
805146           2.0  
805765           2.0  
806119           6.0  

[1910 rows x 18 columns]

In [ ]:
patent_original_data['cpc_2'].unique()

array(['F23', 'B26', 'F41', 'E05', 'B60', 'B21', 'H04', 'F04', 'A01',
       'B68', 'C03', 'C10', 'H03', 'G01', 'E21', 'F42', 'B41', 'F27',
       'Y10', 'E01', 'B22', 'F16', 'B61', 'H01', 'H02', 'B63', 'C21',
       'F25', 'B64', 'B24', 'B25', 'F02', 'F21', 'G05', 'H05', 'B23',
       'C12', 'F01', 'D06', 'E02', 'C08', 'G03', 'D21', 'B65', 'C22',
       'G09', 'B66', nan, 'B29', 'B62', 'B05', 'G11', 'F22', 'C01', 'F28',
       'C25', 'B01', 'B27', 'G08', 'A61', 'A47', 'B42', 'G10', 'A42',
       'C07', 'G06', 'Y02', 'A43', 'A23', 'D04', 'A45', 'B07', 'C13',
       'F03', 'D07', 'F24', 'A21', 'E04', 'C09', 'B30', 'D05', 'C04',
       'D01', 'B32', 'B67', 'E06', 'B02', 'C06', 'F15', 'B28', 'F26',
       'B08', 'G04', 'B04', 'B43', 'E03', 'A41', 'G07', 'C05', 'A24',
       'A63', 'C11', 'C14', 'C02', 'B31', 'G02', 'F05', 'C23', 'G21',
       'D02', 'A22', 'B03', 'A44', 'A62', 'B44', 'A46', 'B06', 'D03',
       'Y04', 'F17', 'C30', 'D10', 'G12', 'B82', 'B09', 'C40', 'B81',
       'G16', '

### 存在空值！！！

In [ ]:
patent_original_data['cpc_g2'].unique()

array([ 99.,  30., 105.,  86.,  42.,  25., 125.,  92.,   1.,  50.,  55.,
        62., 124., 107.,  88., 106.,  38., 103., 129.,  82.,  26.,  95.,
        43., 122., 123.,  45.,  67., 101.,  46.,  28.,  29.,  90.,  97.,
       111., 126.,  27.,  64.,  89.,  78.,  83.,  60., 109.,  81.,  47.,
        68., 115.,  48.,  nan,  33.,  44.,  20., 117.,  98.,  53., 104.,
        70.,  16.,  31., 114.,  13.,  12.,  39., 116.,   7.,  59., 112.,
       127.,   8.,   4.,  76.,  10.,  22.,  65.,  91.,  79., 100.,   2.,
        85.,  61.,  34.,  77.,  56.,  73.,  36.,  49.,  87.,  17.,  58.,
        94.,  32., 102.,  23., 110.,  19.,  40.,  84.,   6., 113.,  57.,
         5.,  15.,  63.,  66.,  54.,  35., 108.,  93.,  69., 120.,  74.,
         3.,  18.,   9.,  14.,  41.,  11.,  21.,  75., 128.,  96.,  71.,
        80., 118.,  52.,  24.,  72.,  51., 119.,  37., 121.])

In [ ]:
patent_original_data['cpc_g2'].max()

129.0

In [ ]:
patent_original_data['cpc_g3'].max()

667.0

In [ ]:
patent_original_data['cpc_g3'].min()

1.0

### cpc_g2中存在NAN空值！！！

In [ ]:
len(patent_original_data['cpc_2'].unique())

130

In [ ]:
len(patent_original_data['cpc_g2'].unique())

130

In [ ]:
patent_original_data['cpc_3'].unique()

array(['F23K', 'B26D', 'F41A', 'E05F', 'B60P', 'B21D', 'H04M', 'H04W',
       'H04B', 'H04J', 'F04B', 'A01D', 'B68G', 'C03C', 'C10G', 'H03F',
       'H04L', 'G01R', 'E21B', 'F42C', 'B41J', 'F27B', 'Y10S', 'H04K',
       'E01B', 'B22C', 'B21B', 'Y10T', 'G01B', 'F16D', 'B61G', 'B60T',
       'C03B', 'B61F', 'H01H', 'H02H', 'H01G', 'B63B', 'H03B', 'B60Q',
       'C21D', 'B60R', 'G01F', 'G01H', 'B60G', 'H04R', 'H02M', 'F25J',
       'H01J', 'B64D', 'B24B', 'B25D', 'F02F', 'F02M', 'G01M', 'F21W',
       'H02K', 'F27D', 'C10L', 'B61C', 'H02J', 'G05F', 'H05B', 'B23K',
       'C12N', 'B60J', 'F01D', 'H04Q', 'F16K', 'D06F', 'E02F', 'B61L',
       'C08K', 'B41K', 'G03B', 'D21F', 'C08B', 'B65B', 'F16C', 'C22C',
       'G09F', 'B66D', 'G01G', 'H01K', 'B65G', 'B23G', 'F23D', nan,
       'H01F', 'B61H', 'F21V', 'B60L', 'B29C', 'H01B', 'G01L', 'B62D',
       'B60M', 'G01K', 'B23B', 'B05C', 'C22B', 'F25B', 'B60B', 'H03C',
       'G11B', 'F01L', 'H03H', 'B62B', 'B29D', 'H04N', 'F22D', 'C01F',
       'F

In [ ]:
len(patent_original_data['cpc_g3'].unique())

668

In [ ]:
patent_original_data['cpc_g3'].unique()

array([480., 139., 508., 406., 201., 109., 644., 650., 639., 641., 437.,
         3., 251., 264., 302., 632., 643., 531., 411., 516., 172., 500.,
       666., 642., 384., 116., 107., 667., 518., 449., 213., 205., 263.,
       212., 610., 622., 609., 226., 629., 202., 330., 203., 521., 523.,
       194., 647., 625., 498., 611., 233., 128., 133., 426., 429., 527.,
       469., 624., 501., 306., 210., 623., 553., 651., 125., 319., 196.,
       417., 646., 454., 366., 392., 217., 289., 173., 539., 380., 283.,
       236., 448., 332., 578., 244., 522., 612., 240., 123., 476.,  nan,
       608., 214., 468., 198., 156., 606., 526., 220., 199., 525., 119.,
        99., 331., 495., 190., 630., 588., 419., 634., 218., 157., 645.,
       472., 259., 422., 423., 528., 503., 245., 241., 340.,  87., 145.,
       529., 191., 619., 541., 572., 148., 633., 385., 540., 621., 238.,
       552., 452.,  65.,  58., 180., 174., 412., 256., 640., 586., 627.,
       617., 656., 573., 170.,  39., 286., 276., 40

In [ ]:
# 筛选出permno=# 筛选出'cpc_g3'为空值的数据
cpc_g3_nan = patent_original_data[patent_original_data['cpc_g3'].isnull()]
cpc_g3_nan

patent_num  cpc cpc_1 cpc_2  cpc_g1  cpc_g2 cpc_3  cpc_g3  permno  \
165         1572782  NaN   NaN   NaN     NaN     NaN   NaN     NaN   10401   
225         1573315  NaN   NaN   NaN     NaN     NaN   NaN     NaN   10401   
514         1577670  NaN   NaN   NaN     NaN     NaN   NaN     NaN   14613   
758         1580392  NaN   NaN   NaN     NaN     NaN   NaN     NaN   15368   
1069        1584748  NaN   NaN   NaN     NaN     NaN   NaN     NaN   10401   
...             ...  ...   ...   ...     ...     ...   ...     ...     ...   
2696449     9850830  NaN   NaN   NaN     NaN     NaN   NaN     NaN   12369   
2696792     9852002  NaN   NaN   NaN     NaN     NaN   NaN     NaN   59328   
2697170     9852930  NaN   NaN   NaN     NaN     NaN   NaN     NaN   85035   
2697236     9853077  NaN   NaN   NaN     NaN     NaN   NaN     NaN   51131   
2697703     9854244  NaN   NaN   NaN     NaN     NaN   NaN     NaN   51131   

         issue_date  filing_date  xi_nominal   xi_real  cites  year_issue  \
165        19260209   19231008.0    0.171491  0.945621      0        1926   
225        19260216   19220912.0    0.178419  0.983825      3        1926   
514        19260323   19240415.0    0.185627  1.023573      1        1926   
758        19260413   19240116.0    0.030381  0.167523      2        1926   
1069       19260518   19191010.0    0.695776  3.836594      1        1926   
...             ...          ...         ...       ...    ...         ...   
2696449    20171226   20141208.0    8.539674  3.380897      0        2017   
2696792    20171226   20150327.0   14.000236  5.542758      0        2017   
2697170    20171226   20161006.0    6.904639  2.733579      0        2017   
2697236    20171226   20170303.0    0.557691  0.220792      0        2017   
2697703    20171226   20170606.0    0.557691  0.220792      0        2017   

         month_issue  year_filing  month_filing  
165                2       1923.0          10.0  
225                2       1922.0           9.0  
514                3       1924.0           4.0  
758                4       1924.0           1.0  
1069               5       1919.0          10.0  
...              ...          ...           ...  
2696449           12       2014.0          12.0  
2696792           12       2015.0           3.0  
2697170           12       2016.0          10.0  
2697236           12       2017.0           3.0  
2697703           12       2017.0           6.0  

[3260 rows x 18 columns]

### 生成相似度邻接矩阵前应该先去除空值！

In [ ]:
# 检查patent_original_data['cpc_g3'].unique()中的数据是否有空值
c = patent_original_data['cpc_g3'].unique().tolist()
# 检查是否有空值，并打印出位置
for i in range(len(c)):
    if pd.isnull(c[i]):
        print(i)

87


In [ ]:
len(patent_original_data['cpc_g3'].unique())

668

In [33]:
patent_original_data['permno'].unique()

array([15368, 11965, 10786, ..., 11275, 14928, 14889])

In [34]:
len(patent_original_data['permno'].unique())

8547

In [35]:
# 保存patent_original_data['permno']
patent_original_data_permno = patent_original_data['permno'].unique().tolist()
patent_original_data_permno

[15368,
 11965,
 10786,
 10807,
 14613,
 10401,
 14736,
 11068,
 14402,
 12431,
 14525,
 15480,
 10751,
 15341,
 13661,
 14699,
 12378,
 12079,
 14090,
 14605,
 10102,
 11164,
 15464,
 12060,
 14533,
 11682,
 14875,
 13119,
 11754,
 10292,
 14795,
 13063,
 10241,
 12546,
 10671,
 15421,
 10057,
 15149,
 10735,
 14429,
 13274,
 14541,
 12140,
 15325,
 10698,
 12503,
 11762,
 13389,
 11957,
 10495,
 10153,
 12095,
 12685,
 13784,
 15202,
 14955,
 10217,
 15528,
 11295,
 10620,
 12263,
 13098,
 10073,
 12597,
 13303,
 14656,
 14912,
 11703,
 11746,
 14322,
 14453,
 13944,
 15552,
 10487,
 10524,
 10188,
 13805,
 10460,
 10006,
 13557,
 14066,
 15376,
 11770,
 10890,
 10030,
 11607,
 13637,
 12466,
 11973,
 15034,
 11850,
 14891,
 11869,
 10778,
 11260,
 10145,
 10161,
 11447,
 15472,
 19334,
 14277,
 15659,
 12087,
 10794,
 11885,
 11826,
 10604,
 10233,
 10372,
 13610,
 11375,
 10874,
 12618,
 14963,
 15253,
 14760,
 12052,
 14701,
 14007,
 12036,
 15683,
 18227,
 13450,
 10022,
 15720,


In [36]:
len(patent_original_data_permno)

8547

In [ ]:
# 筛选出issue year为1980年的数据
patent_original_data_1980 = patent_original_data[patent_original_data['year_issue'] == 1980]

In [ ]:
patent_original_data_1975_1980 = patent_original_data[(patent_original_data['year_issue'] >=1975) & (patent_original_data['year_issue'] < 1980)]
patent_original_data_1975_1980.shape

(121511, 18)

In [ ]:
patent_original_data_1975_1980['permno'].unique()

array([26366, 14656, 12060, ..., 45612, 19887, 63589])

In [ ]:
companies_1975_1980 = patent_original_data_1975_1980['permno'].unique().tolist()

In [ ]:
len(companies_1975_1980)

1853

In [ ]:
patent_original_data_1980.shape

(20773, 18)

In [ ]:
patent_original_data_1980['permno'].unique()

array([19393, 12060, 11607, ..., 58596, 32030, 68320])

In [ ]:
companies_1980 = patent_original_data_1980['permno'].unique().tolist()

In [ ]:
len(companies_1980)

20773

In [ ]:
companies_1980.sort()

In [ ]:
companies_1980

[10006,
 10057,
 10145,
 10153,
 10161,
 10162,
 10189,
 10225,
 10241,
 10321,
 10364,
 10372,
 10401,
 10460,
 10487,
 10488,
 10604,
 10664,
 10672,
 10751,
 10786,
 10808,
 10874,
 10890,
 10989,
 11105,
 11156,
 11164,
 11260,
 11295,
 11296,
 11308,
 11340,
 11447,
 11471,
 11607,
 11703,
 11754,
 11762,
 11850,
 11851,
 12036,
 12044,
 12052,
 12060,
 12079,
 12087,
 12095,
 12140,
 12184,
 12320,
 12431,
 12458,
 12490,
 12503,
 12511,
 12546,
 12570,
 12706,
 12837,
 12918,
 13119,
 13290,
 13303,
 13354,
 13398,
 13557,
 13610,
 13661,
 13689,
 13856,
 13901,
 13928,
 14066,
 14090,
 14189,
 14198,
 14218,
 14227,
 14250,
 14277,
 14286,
 14322,
 14357,
 14525,
 14533,
 14541,
 14592,
 14605,
 14656,
 14736,
 14760,
 14795,
 14824,
 14859,
 14883,
 14891,
 14912,
 14955,
 15034,
 15069,
 15123,
 15202,
 15203,
 15229,
 15368,
 15560,
 15579,
 15659,
 15667,
 15676,
 15720,
 15721,
 15747,
 15763,
 15836,
 15966,
 16011,
 16029,
 16088,
 16109,
 16141,
 16336,
 16360,
 16361,


In [ ]:
# 查看不重复元素的数量
unique_elements = len(set(companies_1980))
print(f"Unique elements: {unique_elements}")

Unique elements: 1033


In [ ]:
len(companies_1980)

1033

In [ ]:
companies_1980[1000:]

[73139,
 73219,
 73649,
 73796,
 74043,
 74617,
 74932,
 75039,
 75064,
 75083,
 75460,
 75540,
 75855,
 75970,
 76655,
 77295,
 77711,
 77762,
 78298,
 78749,
 78800,
 79250,
 79696,
 79901,
 80120,
 80945,
 81382,
 81518,
 81833,
 82262,
 82721,
 82959,
 83433]

存在一个问题：1980.csv文件中 1980年有953家公司的数据，但是RData中的公司数量有1853个，patent_cpc_data.csv文件中1980年的数据中却只有1033家公司

In [ ]:
# 筛选出issue year为1981年的数据
patent_original_data_1981 = patent_original_data[patent_original_data['year_issue'] == 1981]

In [ ]:
companies_1981 = patent_original_data_1981['permno'].unique().tolist()

In [ ]:
len(companies_1981)

1041

In [ ]:
companies_1981.sort()
companies_1981

[10006,
 10057,
 10145,
 10153,
 10161,
 10189,
 10225,
 10241,
 10321,
 10364,
 10372,
 10401,
 10460,
 10479,
 10487,
 10488,
 10516,
 10604,
 10640,
 10656,
 10672,
 10698,
 10751,
 10786,
 10866,
 10874,
 10890,
 10989,
 11105,
 11164,
 11260,
 11308,
 11340,
 11404,
 11447,
 11471,
 11607,
 11674,
 11703,
 11754,
 11762,
 11850,
 11851,
 12036,
 12045,
 12052,
 12060,
 12079,
 12087,
 12095,
 12140,
 12184,
 12431,
 12458,
 12490,
 12503,
 12511,
 12546,
 12570,
 12706,
 12918,
 13056,
 13119,
 13179,
 13303,
 13354,
 13398,
 13557,
 13610,
 13661,
 13689,
 13856,
 13901,
 13928,
 14024,
 14090,
 14189,
 14198,
 14218,
 14227,
 14277,
 14322,
 14323,
 14357,
 14525,
 14533,
 14541,
 14592,
 14593,
 14605,
 14656,
 14736,
 14760,
 14883,
 14891,
 14912,
 15034,
 15069,
 15202,
 15203,
 15229,
 15368,
 15579,
 15659,
 15667,
 15720,
 15721,
 15747,
 15763,
 15828,
 15836,
 15966,
 16029,
 16088,
 16109,
 16141,
 16336,
 16360,
 16361,
 16424,
 16432,
 16433,
 16468,
 16571,
 16598,


In [ ]:
companies_1981[1000:]

[74043,
 74481,
 74908,
 74932,
 75039,
 75062,
 75064,
 75257,
 75329,
 75460,
 75540,
 75970,
 76153,
 76305,
 76655,
 76807,
 76911,
 77295,
 78116,
 78298,
 78749,
 79250,
 79696,
 79901,
 80013,
 80120,
 80128,
 80451,
 80486,
 81139,
 81382,
 81518,
 81700,
 81833,
 82262,
 82721,
 82959,
 83433,
 83687,
 83863,
 84233]

In [ ]:
# 读取RData文件
robjects.r['load']('/home/bld/bld_data/pengliu/Graph data/Patent relation/patent_kg/cor_1980.RData')
patent_Rdata = robjects.r['cor']

In [ ]:
patent_Rdata

<rpy2.robjects.vectors.FloatMatrix object at 0x7fc1068caf10> [14]
R classes: ('matrix', 'array')
[1.000000, 0.246089, 0.232862, 0.000000, ..., 0.000000, 0.000000, 0.000000, 1.000000]

[Running] Rscript "/home/bld/bld_data/pengliu/Graph data/Patent relation/Code/test.r"

Year 1980: NA count = 0, Dimensions =  1853 1853 

Year 1981: NA count = 0, Dimensions =  1831 1831 

Year 1982: NA count = 0, Dimensions =  1819 1819 

Year 1983: NA count = 0, Dimensions =  1792 1792 

Year 1984: NA count = 0, Dimensions =  1792 1792 

Year 1985: NA count = 0, Dimensions =  1847 1847 

Year 1986: NA count = 0, Dimensions =  1870 1870 

Year 1987: NA count = 0, Dimensions =  1936 1936

Year 1988: NA count = 0, Dimensions =  2058 2058 

Year 1989: NA count = 0, Dimensions =  2107 2107 

Year 1990: NA count = 0, Dimensions =  2126 2126 

Year 1991: NA count = 0, Dimensions =  2135 2135 

Year 1992: NA count = 0, Dimensions =  2122 2122 

Year 1993: NA count = 0, Dimensions =  2143 2143 

Year 1994: NA count = 0, Dimensions =  2141 2141 

Year 1995: NA count = 0, Dimensions =  2234 2234 

Year 1996: NA count = 0, Dimensions =  2323 2323 

Year 1997: NA count = 0, Dimensions =  2518 2518 

Year 1998: NA count = 0, Dimensions =  2696 2696 

Year 1999: NA count = 0, Dimensions =  2900 2900 

Year 2000: NA count = 0, Dimensions =  3020 3020 

Year 2001: NA count = 0, Dimensions =  3210 3210 

Year 2002: NA count = 0, Dimensions =  3302 3302 

Year 2003: NA count = 0, Dimensions =  3287 3287 

Year 2004: NA count = 0, Dimensions =  3138 3138 

Year 2005: NA count = 0, Dimensions =  2927 2927 

Year 2006: NA count = 0, Dimensions =  2779 2779 

Year 2007: NA count = 0, Dimensions =  2639 2639 

Year 2008: NA count = 0, Dimensions =  2536 2536 

Year 2009: NA count = 0, Dimensions =  2468 2468 

Year 2010: NA count = 4775, Dimensions =  2388 2388 

Year 2011: NA count = 4647, Dimensions =  2324 2324 

Year 2012: NA count = 4411, Dimensions =  2206 2206 

Year 2013: NA count = 4231, Dimensions =  2116 2116 

Year 2014: NA count = 4123, Dimensions =  2062 2062 

Year 2015: NA count = 0, Dimensions =  2058 2058 

Year 2016: NA count = 0, Dimensions =  1984 1984 

Year 2017: NA count = 0, Dimensions =  1912 1912 

Year 2018: NA count = 0, Dimensions =  1791 1791 

Year 2019: NA count = 0, Dimensions =  1676 1676 

Year 2020: NA count = 0, Dimensions =  1586 1586 

Year 2021: NA count = 0, Dimensions =  1452 1452 

Year 2022: NA count = 0, Dimensions =  1328 1328 